fastlane_bot/utils.py

```diff
@@ -179,15 +183,27 @@ def p_start(self):
     @property
     def p_marg(self):
         if self.y == self.z:
+            # try:
+            new_method = self.decodeRate(self.decodeFloat(int(self.B)) + self.decodeFloat(int(self.A)))
+            # except:
+            #     print(self.B, self.A)
+            #     print(type(self.B), type(self.A))
+            #     print(self.decodeFloat(int(self.B)) + self.decodeFloat(int(self.A)))
+            # print(new_method, self.p_start)
+            assert new_method == self.p_start, f"{new_method}, {self.p_start} **************************************"
             return self.p_start
         elif self.y == 0:
             return self.p_end
-        raise NotImplementedError("p_marg not implemented for non-full / empty orders")
-        A = self.decodeFloat(self.A)
-        B = self.decodeFloat(self.B)
-        return self.decode(B + A * self.y / self.z) ** 2
-        # https://github.com/bancorprotocol/carbon-simulator/blob/beta/benchmark/core/trade/impl.py
-        # 'marginalRate' : decodeRate(B + A if y == z else B + A * y / z),
+        # return 0
+        else:
+            return self.decodeRate(self.decodeFloat(int(self.B)) + (self.decodeFloat(int(self.A)) * self.y/self.z))
```

Comment:
And for clarity, you can compute the following before doing the if/else:
- `A = self.decodeFloat(int(self.A))`
- `B = self.decodeFloat(int(self.B))`

Reply:
added